### PR TITLE
New camera2

### PR DESCRIPTION
**Pull request by TomasAdomavicius**
camera2 implementation

### CODE REVIEW

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+                    var map = (StreamConfigurationMap)characteristics.Get(CameraCharacteristics.ScalerStreamConfigurationMap);
+                    if (map == null)
+                    {
+                        continue;
+                    }
+
+                    // For still image captures, we use the largest available size.
+                    Size largest = (Size)Collections.Max(Arrays.AsList(map.GetOutputSizes((int)ImageFormatType.Jpeg)),
+                        new CompareSizesByArea());
+                    mImageReader = ImageReader.NewInstance(largest.Width, largest.Height, ImageFormatType.Jpeg, /*maxImages*/100);
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs]:
```suggestion
                    mImageReader = ImageReader.NewInstance(largest.Width, largest.Height, ImageFormatType.Jpeg, /*maxImages*/10);
```

File: Mobile/ShopLens/Droid/Activities/MainActivity.cs [hunk 2]
```diff
@@ -2,14 +2,18 @@
 using Android.App;
 using Android.Content;
 using Android.OS;
+using Android.Runtime;
 using Android.Content.PM;
+
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/MainActivity.cs:7]: I can see that you added a new line here.

File: Mobile/ShopLens/Droid/Activities/MainActivity.cs [hunk 3]
```diff
@@ -31,7 +37,8 @@ public enum IntentIds
 
 namespace ShopLens.Droid
 {
-    [Activity(Label = "ShopLens", MainLauncher = true, Icon = "@mipmap/icon", Theme = "@style/ShopLensTheme")]
+    [Activity(Label = "ShopLens", Icon = "@mipmap/icon", Theme ="@style/ShopLensTheme", 
+        ScreenOrientation = ScreenOrientation.Portrait)]
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/MainActivity.cs:41]: No more orienting activities ! 👍

File: Mobile/ShopLens/Droid/Activities/MainActivity.cs [hunk 4]
```diff
@@ -116,12 +122,14 @@ protected override void OnCreate(Bundle savedInstanceState)
...
+            SetContentView(Resource.Layout.Main);            
+
+            
 
-            // We need to request user permissions.
-            if ((int)Build.VERSION.SdkInt >= (int)BuildVersionCodes.M)
+            if (savedInstanceState == null)
             {
-                RequestPermissions(ShopLensPermissions, REQUEST_PERMISSION);
+                new Thread(() => { FragmentManager.BeginTransaction().Replace(Resource.Id.container, Camera2Fragment.NewInstance(this, this)).Commit(); }
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/MainActivity.cs:131]: I've tested this without the new Thread creation and it works fine for me, so test this on your device and find out whether a new Thread is really necessary.

File: Mobile/ShopLens/Droid/Activities/PermissionActivity.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
+﻿using Android;
+using Android.App;
+using Android.Content;
+using Android.Content.PM;
+using Android.OS;
+
+namespace ShopLens.Droid.Activities
+{
+    [Activity(Label = "PermissionActivity", MainLauncher = true)]
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/PermissionActivity.cs:9]: I think this activity should have the label "ShopLens" and it should use our app icon, also its orientation should be portrait and it would also be nice if it wasn't dark but white in color when launched.

File: Mobile/ShopLens/Droid/Activities/MainActivity.cs [hunk 6]
```diff
@@ -55,32 +70,23 @@ public class MainActivity : AppCompatActivity
...
         public static bool goingFromCartToList = false;
         public static bool goingFromListToCart = false;
         public static List<string> shoppingSessionItems;
 
-        readonly string[] ShopLensPermissions =
+        string needUserAnswerId;
+        string askUserToRepeat;
+        string talkBackEnabledIntentKey;
+
+        public readonly string[] ShopLensPermissions =
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/MainActivity.cs:83]: Should this unused variable be in this class, that doesn't use it at all?

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs [hunk 7]
```diff
@@ -11,13 +11,18 @@
...
+using Android.Support.V7.App;
 using Android.Views;
 using ShopLens.Droid.Activities;
 
 namespace ShopLens.Droid
 {
-    [Activity(Label = "ShoppingCartActivity", Theme = "@style/ShopLensTheme")]
-    public class ShoppingCartActivity : Activity
+    // TODO: Make voice interaction in shopping cart fluent.
+    [Activity(Label = "Shopping Cart", Theme = "@style/ShopLensTheme")]
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs:25]: I think that TODO ir already done, so it can be deleted.

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs [hunk 8]
```diff
@@ -11,13 +11,18 @@
...
+using SupportToolbar = Android.Support.V7.Widget.Toolbar;
+using Android.Support.V7.App;
 using Android.Views;
 using ShopLens.Droid.Activities;
 
 namespace ShopLens.Droid
 {
-    [Activity(Label = "ShoppingCartActivity", Theme = "@style/ShopLensTheme")]
-    public class ShoppingCartActivity : Activity
+    // TODO: Make voice interaction in shopping cart fluent.
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs:24]:
```suggestion
```

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs (hunk 7 above)
Comment (marjasi) [reply Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs:25]: I also think that this activity should have its orientation set properly.

File: Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs [hunk 9]
```diff
@@ -6,22 +6,32 @@
...
 using ShopLens.Droid.Helpers;
 using ShopLens.Droid.Source;
 using ShopLens.Extensions;
+using SupportToolbar = Android.Support.V7.Widget.Toolbar;
 
 namespace ShopLens.Droid
 {
-    [Activity(Label = "ShoppingListActivity", Theme = "@style/ShopLensTheme")]
-    public class ShoppingListActivity : Activity
+    [Activity(Label = "Shopping List", Theme = "@style/ShopLensTheme")]
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs:25]: I think that this activity should have its orientation set properly.

File: Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs [hunk 10]
```diff
@@ -84,8 +94,47 @@ protected override void OnCreate(Bundle savedInstanceState)
             listView.ChoiceMode = ChoiceMode.Multiple;
 
             addItemButton.Click += AddTextBoxProductToList;
-            removeItemButton.Click += RemoveTextBoxProductFromList;
-            removeAllItemsButton.Click += RemoveAllItems;
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs]: Why is this deleted? Now we can't remove items with the buttons?

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs [hunk 11]
```diff
@@ -84,8 +96,48 @@ protected override void OnCreate(Bundle savedInstanceState)
...
+            drawerToggle.SyncState();
+
+            navView.NavigationItemSelected += (sender, e) =>
+            {
+                switch (e.MenuItem.ItemId)
+                {
+                    case Resource.Id.NavItemShoppingCart:
+                        OnOptionsItemSelected(e.MenuItem);
+                        break;
+                    case Resource.Id.NavItemShoppingList:
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs:127]: This is cool, but you can't go back to the main menu with it.

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs [hunk 12]
```diff
@@ -84,8 +96,48 @@ protected override void OnCreate(Bundle savedInstanceState)
...
+            SetSupportActionBar(toolbar);
+            SupportActionBar.SetDisplayHomeAsUpEnabled(true);
+            SupportActionBar.SetHomeButtonEnabled(true);
+            drawerToggle.SyncState();
+
+            navView.NavigationItemSelected += (sender, e) =>
+            {
+                switch (e.MenuItem.ItemId)
+                {
+                    case Resource.Id.NavItemShoppingCart:
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs:124]: You're in shopping cart, so why would you want to go to shopping cart? Why is this here?

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs [hunk 13]
```diff
@@ -84,8 +96,48 @@ protected override void OnCreate(Bundle savedInstanceState)
             listView.ChoiceMode = ChoiceMode.Multiple;
 
             addItemButton.Click += AddTextBoxProductToList;
-            removeItemButton.Click += RemoveTextBoxProductFromList;
-            removeAllItemsButton.Click += RemoveAllItems;
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs]: I can't remove items from my shopping cart, please help me, the voice commands don't work half the time, what will I do?

File: Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs [hunk 14]
```diff
@@ -84,8 +94,47 @@ protected override void OnCreate(Bundle savedInstanceState)
...
+            drawerToggle.SyncState();
+
+            navView.NavigationItemSelected += (sender, e) =>
+            {
+                switch (e.MenuItem.ItemId)
+                {
+                    case Resource.Id.NavItemShoppingCart:
+                        StartCartIntent();
+                        break;
+                    case Resource.Id.NavItemShoppingList:
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/ShoppingListActivity.cs:123]: This shouldn't be visible in the drawer, also no way to go back to main menu from here.

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 15]
```diff
@@ -0,0 +1,550 @@
...
+                for (var i = 0; i < manager.GetCameraIdList().Length; i++)
+                {
+                    var cameraId = manager.GetCameraIdList()[i];
+                    CameraCharacteristics characteristics = manager.GetCameraCharacteristics(cameraId);
+
+                    // We don't use a front facing camera.
+                    var facing = (Integer)characteristics.Get(CameraCharacteristics.LensFacing);
+                    if (facing != null && facing == (Integer.ValueOf((int)LensFacing.Front)))
+                    {
+                        continue;
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs:187]: Never seen this one before.

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 16]
```diff
@@ -0,0 +1,550 @@
...
+                        mFlashSupported = (bool)available;
+                    }
+
+                    mCameraId = cameraId;
+                    return;
+                }
+            }
+            catch (CameraAccessException e)
+            {
+                e.PrintStackTrace();
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs:245]: I think Aseris will not like this exception handling. You either make a custom exception and save this exception to it and throw the custom one or do something meaningful about this exception.

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 17]
```diff
@@ -0,0 +1,550 @@
...
+            }
+            catch (NullPointerException)
+            {
+                // Currently an NPE is thrown when the Camera2API is used but not supported on the
+                // device this code runs.
+                ErrorDialogCreator error = new ErrorDialogCreator(Activity.ApplicationContext, "title", "This device doesn\'t support Camera2 API.", "OK");
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs:251]: Now this is way better! Maybe you should handle CameraAccessException the same way?

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 18]
```diff
@@ -0,0 +1,550 @@
...
+
+        // Opens the camera specified by {@link Camera2Fragment#mCameraId}.
+        public void OpenCamera(int width, int height)
+        {
+            SetUpCameraOutputs(width, height);
+            var activity = Activity;
+            var manager = (CameraManager)activity.GetSystemService(Context.CameraService);
+            try
+            {
+                if (!mCameraOpenCloseLock.TryAcquire(2500, TimeUnit.Milliseconds))
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs:264]: 2500 seems like a random number, Aseris won't like that.

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 19]
```diff
@@ -0,0 +1,550 @@
...
+        }
+
+        // Sets up member variables related to camera.
+        private void SetUpCameraOutputs(int width, int height)
+        {
+            var activity = Activity;
+            var manager = (CameraManager)activity.GetSystemService(Context.CameraService);
+            try
+            {
+                for (var i = 0; i < manager.GetCameraIdList().Length; i++)
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs:178]: This is a good place to use a foreach loop. Use foreach.

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 20]
```diff
@@ -0,0 +1,550 @@
...
+                }
+                manager.OpenCamera(mCameraId, mStateCallback, mBackgroundHandler);
+            }
+            catch (CameraAccessException e)
+            {
+                e.PrintStackTrace();
+            }
+            catch (InterruptedException e)
+            {
+                throw new RuntimeException("Interrupted while trying to lock camera opening.", e);
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs:276]: Do you see that this exception is being thrown and it is passed a custom message to be informative and also the original InterruptedException? That's how you should handle exceptions in a custom way.

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 21]
```diff
@@ -0,0 +1,550 @@
...
+            {
+                if (!mCameraOpenCloseLock.TryAcquire(2500, TimeUnit.Milliseconds))
+                {
+                    throw new RuntimeException("Time out waiting to lock camera opening.");
+                }
+                manager.OpenCamera(mCameraId, mStateCallback, mBackgroundHandler);
+            }
+            catch (CameraAccessException e)
+            {
+                e.PrintStackTrace();
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs:272]: No error dialog or exception throw?

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 22]
```diff
@@ -0,0 +1,550 @@
...
+            mBackgroundThread.QuitSafely();
+            try
+            {
+                mBackgroundThread.Join();
+                mBackgroundThread = null;
+                mBackgroundHandler = null;
+            }
+            catch (InterruptedException e)
+            {
+                e.PrintStackTrace();
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs:332]: Error dialog or maybe a throw?

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 23]
```diff
@@ -0,0 +1,550 @@
...
+                {
+                    surface,
+                    mImageReader.Surface
+                };
+                mCameraDevice.CreateCaptureSession(surfaces, new CaptureSessionCallback(this), null);
+
+            }
+            catch (CameraAccessException e)
+            {
+                e.PrintStackTrace();
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs:368]: I think you should make your custom exception for CameraAccessException, because this exception is catched so many times.

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 24]
```diff
@@ -0,0 +1,550 @@
...
+                SetAutoFlash(stillCaptureBuilder);
+
+                stillCaptureBuilder.Set(CaptureRequest.JpegOrientation, 90);//TODO
+
+                mCaptureSession.StopRepeating();
+                mCaptureSession.Capture(stillCaptureBuilder.Build(), new CaptureStillPictureSessionCallback(this), null);
+            }
+            catch (CameraAccessException e)
+            {
+                e.PrintStackTrace();
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs:444]: Again

File: Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs [hunk 25]
```diff
@@ -0,0 +1,550 @@
...
+                mCaptureSession.Capture(mPreviewRequestBuilder.Build(), mCaptureCallback,
+                        mBackgroundHandler);
+                // After this, the camera will go back to the normal state of preview.
+                mState = STATE_PREVIEW;
+                mCaptureSession.SetRepeatingRequest(mPreviewRequest, mCaptureCallback,
+                        mBackgroundHandler);
+            }
+            catch (CameraAccessException e)
+            {
+                e.PrintStackTrace();
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Fragments/Camera2Fragment.cs:466]: And again

File: Mobile/ShopLens/Droid/Helpers/ImageRecognizer.cs [hunk 26]
```diff
@@ -0,0 +1,169 @@
...
+using Unity;
+using PCLAppConfig;
+using Android.App;
+using ImageRecognitionMobile.OCR;
+
+namespace ShopLens.Droid
+{
+    class ImageRecognizer
+    {
+        TextToSpeech tts;
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Helpers/ImageRecognizer.cs:27]: If you're using ShopLensTts, why is this here? Will it be used in this class?

File: Mobile/ShopLens/Droid/Helpers/ImageRecognizer.cs [hunk 27]
```diff
@@ -0,0 +1,169 @@
...
+        Camera2Fragment owner;
+        private readonly string prefsName = ConfigurationManager.AppSettings["ShopCartPrefs"];
+
+        public ImageRecognizer(Context context, Activity activity)
+        {
+            this.activity = activity;
+            this.context = context;
+            textRecognizer = new TextRecognizer.Builder(context).Build();
+            sltts = new ShopLensTextToSpeech(context, OnTextToSpeechInit, OnTextToSpeechEndOfSpeech);
+            tts = new TextToSpeech(context, sltts);
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Helpers/ImageRecognizer.cs:43]: You don't need to create a tts object when sltts is doing all the talking for you.

File: Mobile/ShopLens/Droid/Helpers/ImageRecognizer.cs [hunk 28]
```diff
@@ -0,0 +1,169 @@
...
+                    //progressBar.Visibility = ViewStates.Gone;
+                    if (task.IsFaulted)
+                    {
+                        System.Diagnostics.Debug.WriteLine(task.Exception);
+                        return;
+                    }
+                    var recognitionResult = task.Result;
+                    var thingsToSay = new List<string>
+                        {$"This is {recognitionResult.BestPrediction}", recognitionResult.WeightSpecifier};
+                    tts.Speak(
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Helpers/ImageRecognizer.cs:99]: Why aren't you using ShopLensTts?

File: Mobile/ShopLens/Droid/Helpers/ImageRecognizer.cs [hunk 29]
```diff
@@ -0,0 +1,169 @@
...
+            var maxWebClassifierImageSize = int.Parse(ConfigurationManager.AppSettings["webClassifierImgSize"]);
+            bitmap = BitmapHelper.ScaleDown(bitmap, maxWebClassifierImageSize);
+            return bitmap;
+        }
+
+        private void AddToShoppingCart(string guess)
+        {
+            var preferences = new ActivityPreferences(context, prefsName);
+            preferences.AddString(guess.FirstCharToUpper());
+            //shoppingCartItemAddedMessageBar.Show();
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Helpers/ImageRecognizer.cs:122]: I think this should either be removed or uncommented.

File: Mobile/ShopLens/Droid/Helpers/ImageRecognizer.cs [hunk 30]
```diff
@@ -0,0 +1,169 @@
...
+using PCLAppConfig;
+using Android.App;
+using ImageRecognitionMobile.OCR;
+
+namespace ShopLens.Droid
+{
+    class ImageRecognizer
+    {
+        TextToSpeech tts;
+        bool isTtsInitialized = false;
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Helpers/ImageRecognizer.cs:28]: Is this value used anywhere else?

File: Mobile/ShopLens/Droid/Listeners/CaptureListener.cs [hunk 31]
```diff
@@ -0,0 +1,81 @@
...
+
+namespace ShopLens.Droid.Listeners
+{
+    public class CaptureListener : CameraCaptureSession.CaptureCallback
+    {
+        private readonly Camera2Fragment owner;
+
+        public CaptureListener(Camera2Fragment owner)
+        {
+            this.owner = owner ?? throw new System.ArgumentNullException("owner");
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Listeners/CaptureListener.cs:12]: Is "owner" descriptive enough?

File: Mobile/ShopLens/Droid/Listeners/CaptureSessionCallback.cs [hunk 32]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace ShopLens.Droid.Listeners
+{
+    public class CaptureSessionCallback : CameraCaptureSession.StateCallback
+    {
+        private readonly Camera2Fragment owner;
+
+        public CaptureSessionCallback(Camera2Fragment owner)
+        {
+            this.owner = owner ?? throw new System.ArgumentNullException("owner");
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Listeners/CaptureSessionCallback.cs:11]: Maybe this is descriptive enough, because it tells you the line at which you get it but still.

File: Mobile/ShopLens/Droid/Listeners/CaptureSessionCallback.cs [hunk 33]
```diff
@@ -0,0 +1,47 @@
...
+                owner.SetAutoFlash(owner.mPreviewRequestBuilder);
+
+                // Finally, we start displaying the camera preview.
+                owner.mPreviewRequest = owner.mPreviewRequestBuilder.Build();
+                owner.mCaptureSession.SetRepeatingRequest(owner.mPreviewRequest,
+                        owner.mCaptureCallback, owner.mBackgroundHandler);
+            }
+            catch (CameraAccessException e)
+            {
+                e.PrintStackTrace();
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Listeners/CaptureSessionCallback.cs:43]: Throw that exception.

File: Mobile/ShopLens/Droid/Listeners/CaptureStillPictureSessionCallback.cs [hunk 34]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace ShopLens.Droid.Listeners
+{
+    public class CaptureStillPictureSessionCallback : CameraCaptureSession.CaptureCallback
+    {
+        private readonly Camera2Fragment owner;
+
+        public CaptureStillPictureSessionCallback(Camera2Fragment owner)
+        {
+            this.owner = owner ?? throw new System.ArgumentNullException("owner");
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Listeners/CaptureStillPictureSessionCallback.cs:11]: "owner"

File: Mobile/ShopLens/Droid/Listeners/ImageAvailableListener.cs [hunk 35]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        Context context;
+        Activity activity;
+
+        public ImageAvailableListener(Activity activity, Context context, Camera2Fragment fragment)
+        {
+            this.activity = activity;
+            this.context = context;
+            if (fragment == null)
+                throw new System.ArgumentNullException("fragment");
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Listeners/ImageAvailableListener.cs:22]: "fragment"?

File: Mobile/ShopLens/Droid/Listeners/StateListener.cs [hunk 36]
```diff
@@ -0,0 +1,46 @@
...
+namespace ShopLens.Droid.Listeners
+{
+    public class StateListener : CameraDevice.StateCallback
+    {
+        private readonly Camera2Fragment owner;
+
+        public StateListener(Camera2Fragment owner)
+        {
+            if (owner == null)
+                throw new System.ArgumentNullException("owner");
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Listeners/StateListener.cs:13]: "owner"

File: Mobile/ShopLens/Droid/Listeners/SurfaceTextureListener.cs [hunk 37]
```diff
@@ -0,0 +1,36 @@
...
+namespace ShopLens.Droid.Listeners
+{
+    public class SurfaceTextureListener : Java.Lang.Object, TextureView.ISurfaceTextureListener
+    {
+        private readonly Camera2Fragment owner;
+
+        public SurfaceTextureListener(Camera2Fragment owner)
+        {
+            if (owner == null)
+                throw new System.ArgumentNullException("owner");
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Listeners/SurfaceTextureListener.cs:12]: "owner"

File: Mobile/ShopLens/Droid/Listeners/SurfaceTextureListener.cs [hunk 38]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        public bool OnSurfaceTextureDestroyed(Android.Graphics.SurfaceTexture surface)
+        {
+            return true;
+        }
+
+        public void OnSurfaceTextureSizeChanged(Android.Graphics.SurfaceTexture surface, int width, int height)
+        {
+
```
Comment (MellowJay) [inline Mobile/ShopLens/Droid/Listeners/SurfaceTextureListener.cs]: Seems really empty.

File: Mobile/ShopLens/Droid/Listeners/SurfaceTextureListener.cs [hunk 39]
```diff
@@ -0,0 +1,36 @@
...
+        }
+
+        public void OnSurfaceTextureSizeChanged(Android.Graphics.SurfaceTexture surface, int width, int height)
+        {
+        
+        }
+
+        public void OnSurfaceTextureUpdated(Android.Graphics.SurfaceTexture surface)
+        {
+
```
Comment (MellowJay) [inline Mobile/ShopLens/Droid/Listeners/SurfaceTextureListener.cs]: Seems really empty.

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs [hunk 40]
```diff
@@ -11,13 +11,19 @@
...
 using ShopLens.Droid.Activities;
+using ActionBarDrawerToggle = Android.Support.V7.App.ActionBarDrawerToggle;
 
 namespace ShopLens.Droid
 {
-    [Activity(Label = "ShoppingCartActivity", Theme = "@style/ShopLensTheme")]
-    public class ShoppingCartActivity : Activity
+    // TODO: Make voice interaction in shopping cart fluent.
+    [Activity(Label = "Shopping Cart", Theme = "@style/ShopLensTheme")]
+    public class ShoppingCartActivity : AppCompatActivity
```
Comment (marjasi) [inline Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs:26]: You forgot about this.
image

**Review by marjasi — COMMENTED** [review]
Hey, good job on the camera2 👍 👍, now we can finally feel that our app is a real "app" 😃. I left comments here and there. Since we don't have a lot of time, I say we merge and then you will have to look into these comments later.

**Comment by marjasi** [comment]
[[Feature] Move Camera activity to Main activity and make camera appear inside app. (Priority - HIGH)](https://trello.com/c/acxhlz9K/81-feature-move-camera-activity-to-main-activity-and-make-camera-appear-inside-app-priority-high)

**Comment by marjasi** [comment]
[[Feature] Delete Camera activity](https://trello.com/c/WFM9YODU/131-feature-delete-camera-activity)

**Comment by marjasi** [comment]
[[Bug] Disable screen rotation. (Priority - Normal)](https://trello.com/c/vwY9O0jE/103-bug-disable-screen-rotation-priority-normal)

**Comment by marjasi** [comment]
[[Feature] Make toolbar usable in all activities. (Priority - HIGH)](https://trello.com/c/1IAb1nsJ/117-feature-make-toolbar-usable-in-all-activities-priority-high)

File: Mobile/ShopLens/Droid/Activities/MainActivity.cs (hunk 4 above)
Comment (TomasAdomavicius) [reply Mobile/ShopLens/Droid/Activities/MainActivity.cs:131]: maybe after testing designer or other auto generated file didn't change therefore you didn't get an error. I think it's better to leave as it is.

File: Mobile/ShopLens/Droid/Activities/MainActivity.cs (hunk 6 above)
Comment (TomasAdomavicius) [reply Mobile/ShopLens/Droid/Activities/MainActivity.cs:83]: actually they are all used, however, someone else created them and I can't say are they really needed.

File: Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs (hunk 7 above)
Comment (TomasAdomavicius) [reply Mobile/ShopLens/Droid/Activities/ShoppingCartActivity.cs:25]: maybe it was merged mistake.

File: Mobile/ShopLens/Droid/Listeners/SurfaceTextureListener.cs (hunk 37 above)
Comment (EdvardasDlugauskas) [reply Mobile/ShopLens/Droid/Listeners/SurfaceTextureListener.cs:12]: Better use `nameof(owner)` instead

File: Mobile/ShopLens/Droid/Listeners/SurfaceTextureListener.cs (hunk 38 above)
Comment (EdvardasDlugauskas) [inline Mobile/ShopLens/Droid/Listeners/SurfaceTextureListener.cs:28]: If the interface requires an implementation but we want to ignore it, leaving the method empty seems reasonable.

Another question is whether we want to ignore it.
